app/javascript/components/student/editor/TestRunSummary.tsx

```diff
@@ -57,37 +48,59 @@ export function TestRunSummary({
   submission: Submission
   timeout: number
 }) {
-  const RESOLVED_TEST_STATUSES = [
-    TestRunStatus.PASS,
-    TestRunStatus.FAIL,
-    TestRunStatus.ERROR,
-    TestRunStatus.OPS_ERROR,
-  ]
-  const [testRun, dispatch] = useReducer(reducer, {
+  const [testRun, setTestRun] = useState<TestRun>({
     submissionUuid: submission.uuid,
     status: submission.testsStatus,
     message: '',
     tests: [],
   })
-  const haveTestsResolved = RESOLVED_TEST_STATUSES.includes(testRun.status)
+  const channel = useRef<TestRunChannel | undefined>()
+  const timer = useRef<number | undefined>()
+  const handleQueued = useCallback(() => {
+    timer.current = window.setTimeout(() => {
+      setTestRun({ ...testRun, status: TestRunStatus.TIMEOUT })
+      timer.current = undefined
+    }, timeout)
+  }, [timer])
+  const handleTimeout = useCallback(() => {
+    channel.current?.disconnect()
+  }, [channel])
+
+  useEffect(() => {
+    switch (testRun.status) {
+      case TestRunStatus.QUEUED:
+        handleQueued()
```

Comment:
Is there a reason these are not just inlined here?

Reply:
@iHiD and @ErikSchierboom suggested this. I don't have any preference. I'd rather you talk it out then I will apply the changes.

Reply:
The level of complexity here before made this method just pretty hard to read. Might be easier for more experienced React devs, but both Erik and I found it too deep to be able to easily follow.